src/critics/local_goal_critic.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 @artofnothingness Alexey Budyakov, Samsung Research
+#include "mppic/critics/local_goal_critic.hpp"
+
+#include <xtensor/xmath.hpp>
+#include <xtensor/xsort.hpp>
+
+namespace mppi::critics
+{
+
+void LocalGoalCritic::initialize()
+{
+  auto getParam = parameters_handler_->getParamGetter(name_);
+
+  getParam(consider_angle_, "consider_angle", true);
+  getParam(angle_offset_, "angle_offset", 6);
+  getParam(angle_cost_power_, "angle_cost_power", 1);
+  getParam(angle_cost_weight_, "angle_cost_weight", 3.0);
+
+  getParam(consider_distance_, "consider_distance", true);
+  getParam(distance_goal_count_, "distance_goal_count", 2);
+  getParam(distance_offset_, "distance_offset", 6);
+  getParam(distance_cost_power_, "distance_cost_power", 2);
+  getParam(distance_cost_weight_, "distance_cost_weight", 3.0);
+
+  getParam(stop_usage_path_reached_ratio_, "stop_usage_path_reached_ratio", 0.35);
+}
+
+void LocalGoalCritic::score(models::CriticFunctionData & data)
+{
+  if(!enabled_) {
+    return;
+  }
+
+  auto path_points = xt::view(data.path, xt::all(), xt::range(0, 2));
+
+  // get furthest point among trajectories nearest points to path
+  size_t furthest_reached_path_point_idx;
+  if (data.furthest_reached_path_point) {
+    furthest_reached_path_point_idx = *data.furthest_reached_path_point;
+  } else {
+    furthest_reached_path_point_idx = findPathFurthestPoint(data);
+  }
+
+  // calculate range of points to consider as local goals
+  auto lower_distance_offset = std::min(furthest_reached_path_point_idx + distance_offset_, path_points.shape(0) - 1);
+  auto upper_distance_offset = std::min(lower_distance_offset + distance_goal_count_, path_points.shape(0) - 1);
+
+  // calculate angle point as local goal
+  auto angle_offset = std::min(
+    furthest_reached_path_point_idx + angle_offset_, path_points.shape(
+      0) - 1);
+
+  // portion of the path which if reached by trajectories then critic returns
+  size_t threshold_idx = static_cast<size_t>(
+    static_cast<double>(data.path.shape(0) - 1) * stop_usage_path_reached_ratio_);
+
+  if (lower_distance_offset < threshold_idx) {
+    return;
+  }
+
+  if (consider_distance_ ) {
```

Comment:
I've spent a great deal of time thinking about this last night. What I **think** the effect of this critic is is essentially following the carrot. Since you're giving it a set of points by their definition that are after the furthest point on the trajectory, its driving the scoring to select trajectories with higher speeds.

Is that correct?

That seems instead the role of something like https://github.com/artofnothingness/mppic/issues/50 so I wonder if this should not be replaced with that (if I am correct). 

---

Separately, why the upper / lower bounds? Lets assume the path is a straight line. If you find the distances between a range of points relative to the last trajectory point, then average them, it would be the same thing as just taking the distance between the mid-point `lower_distance_offset + upper_distance_offset/ 2` and the trajectory. 

If we don't assume a straight line, instead its turning on a curve, I believe that the distance would be depressed (shorter) since the straight line `L2` lengths are going to be shorter than the arc length that the path actually is. 

In some ways, would that not slow, slightly, the trajectories in turns? Do we actually see that behavior with this critic? Again, assuming my commentary in the section above is correct. 

If so, this seems like an odd, but potentially very functional critic that is itself the speed critic. Maybe with that in mind, there's some modifications that can be made to the method to convert it more 'officially' into the speed critic?

Reply:
> I've spent a great deal of time thinking about this last night. What I think the effect of this critic is is essentially following the carrot. Since you're giving it a set of points by their definition that are after the furthest point on the trajectory, its driving the scoring to select trajectories with higher speeds.
Is that correct?

Not higher speed, but closer to path point that is furthest-nearest, i would say, but generally it should push trajectories toward the path.

Reply:
Separately, why the upper / lower bounds?

![image](https://user-images.githubusercontent.com/48126993/168376326-ddc9f5fe-e2d7-4d4c-838f-17f9cec0d45c.png)

Reply:
I want to point out also that only distance part use several points. Angle part use just some offseted point

Reply:
https://webwhiteboard.com/board/MuIrUkergqsVkbXS3X5GQdfQsR8v6phn/
This is temporary link to the board, so you can draw something too

Reply:
So my idea was to pull trajectories toward the path furthest-nearest points with some offset, for distances use range of points for angles just one point.  Angles used so, to turn robot toward some offseted furthest-nearset point. And all of this stuffs must work only **IF** our trajectories not fully spread toward the path

Reply:
PS
In picture i forgot to draw offset to furthest
so it yellow points must be furthest +offset to furthest + offset + count

Reply:
It has some corner cases where angle part really prevents turning. But i think to manage this by turning off angle part if angle between robot heading and furthest-neareset less than some value, but i think in later PR

Reply:
> Not higher speed, but closer to path point that is furthest-nearest,

But if we extrapolate that 1 degree further: how does the trajectory get to be closer to a further-away point? It needs to be faster (e.g. weights faster moving trajectories over many iterations to speed up the robot). Again, I don't think this is a problem, but it is (I believe) an artifact of this since even if we're exactly following the path, this cost does not approach `0`. 

Not talking about the angles stuff (`consider_angle_`) in this thread, just the distance stuff (`consider_distance_`) - I think we're in agreement on the angle stuff and a single point check is correct.

In your diagram you said that you wanted it to be pulled by multiple path points -- why? I don't understand that. 

If for instance we did the path angle stuff in `consider_angle_` and we did the reference path critic, but without this stuff in `consider_distance_`, what would be the behavior of the system? What problem is this bit of code solving?

Reply:
> But if we extrapolate that 1 degree further: how does the trajectory get to be closer to a further-away point? It needs to be faster (e.g. weights faster moving trajectories over many iterations to speed up the robot). Again, I don't think this is a problem, but it is (I believe) an artifact of this since even if we're exactly following the path, this cost does not approach 0. I could be wrong though.

You are right here. It pull trajectories forward as well as goal critic. It could be Interpreted like speed critic, but then we must interpret goal critic similarly. Basically i have in mind that desired speed critic is about scalar value critic, here we have direction.  

> If for instance we did the path angle stuff in consider_angle_ and we did the reference path critic, but without this stuff in consider_distance_, what would be the behavior of the system? What problem is this bit of code solving and does the robot system slow down?

If you mean angle part of this critic plus old reference critic (path align critic for now), then we will have some part that just align trajectory to path, and the part which will turn robot toward local goal if trajectories not fully spread (robot stuck in something so all trajectories near robot itself, then this critic try to turn us to the nearest local goal)

> In your diagram you said that you wanted it to be pulled by multiple path points -- why? I don't understand that. Just less sensitive to choosing a bad single point or is there something more?

Mostly intuition here. The form of made cost field more elongated, so trajectories probably would seek more to follow the path and not gather around one point

Reply:
![image](https://user-images.githubusercontent.com/48126993/168391248-2a382818-0c44-4184-ba6b-6f763e8b1784.png)

Reply:
![image](https://user-images.githubusercontent.com/48126993/168391455-95e73cd2-4e7a-422b-bc68-936ec4c7ab06.png)

Reply:
>  but then we must interpret goal critic similarly

Yeah, I think that's also a fair interpretation. Though, semantically the goal critic actually drives the robot towards the goal, but we already have another critic that drives towards the path and its not immediately clear to me that the `consider_distance_` stuff actually improves path tracking performance or it just drives it forward.  

> Basically i have in mind that desired speed critic is about scalar value critic, here we have direction.

Why a scalar vs direction? This seems to work pretty good, even if it wasn't the outright intent. I'd think tweaking the `consider_distance_` a bit with the intention of making the speed critic would be really clever.  

> If you mean angle part of this critic plus old reference critic (path align critic for now)

If we have `consider_distance_` false, `consider_angle_` true and use the current state of the refactored branch in #52 (which removed this stuff), how well does it fair? What is the contribution of behavior that the logic under `consider_distance_` provides the system (e.g. just speed, improved path tracking, or anti-path wobbling, etc)?

> then we will have some part that just align trajectory to path, and the part which will turn robot toward local goal if trajectories not fully spread (robot stuck in something so all trajectories near robot itself, then this critic try to turn us to the nearest local goal)

Do you mean that the trajectories would just get really short (eg low speed) and cluster around the path? Is there any improvement to path tracking performance or wobbling reduction?

If the main contribution is just spreading out the trajectory points (e.g. going faster), then it seems, at least for now, that this is acting like the speed critic as its main contribution to behavior (if its not improving wobbling or path tracking)

Reply:
> consider_distance_ stuff actually improves path tracking performance or it just drives it forward

It should improve tracking performance of path tracking, especially in case of stucking in high cost fields (obstacles around). It was meant to 1) change direction of following from global goal to global + local if we stuck (not fully spread trajectories) 2) increase the cost to follow path if we stuck (e.g a lot of obstacles around). 

> Do you mean that the trajectories would just get really short (eg low speed) and cluster around the path? Is there any improvement to path tracking performance or wobbling reduction?

I mean that consider_angle_stuff and consider_distance_ will work only if trajectories not in radius of distance_to_goal parameter. I set this parameter so, that critic activates if trajectories get short. (stuck in obstacle, a lot of obstacles around etc). Then if consider_distance_ false, consider_angle_ true this stuff will turn robot to the path, but not increase following cost, and not pull robot to the nearest goal, it will be pulled only by last path point goal, so it  1) couldn't help to overcome high cost space, 2 not help with pulling trajectories to local goal, so offset it more far goal which lead it to short trajectories (stuck or something). It must correct goal critic from stucking

Reply:
![image](https://user-images.githubusercontent.com/48126993/168400079-4d13e43d-2e20-42dc-aea0-d57aae308dae.png)

Reply:
![image](https://user-images.githubusercontent.com/48126993/168400125-258b2bf7-0c21-427e-a2a2-f010ddf092ce.png)

Reply:
![image](https://user-images.githubusercontent.com/48126993/168400272-e4740f85-e5a3-4870-83c8-901ffe01db9d.png)

Reply:
Ah I see, thanks for explaining it like that. I totally get it now. 

> if consider_distance_ false, consider_angle_ true critic will turn robot to the path, but not increase following cost, and not pull robot to the nearest goal, it will be pulled only by last path point goal (Goal Critic)

because the reference path critic (err renamed to `PathAlignCritic`) by itself won't do this. If there's no "pull" it will align but not go along well, so I understand then. This bit of code is strictly required with path tracking, since both of those situations you bring up are pretty common. 

So here are my thoughts on this break out (also covers some of the other comments above):
- if we make the `consider_angle_` --> `PathAngle`, then `PathAngle` becomes usable and we should start to use it to decompose the utility parts of the critic added in this PR
- My opinion is that the `consider_distance_` logic is so fundamental to successful path tracking, that it should be put back into the reference trajectory critic and rename it again `ReferencePathCritic`. That would put all the important stuff in 1 place and prevent users from shooting themselves in the foot by removing critics and getting bad performance.
- However, if you disagree and would rather 2 critics `PathAlign` / `PathFollow`,  I think your name of `PathFollowCritic` is very apt and I can understand the reasoning for wanting them separate. 

So I leave that choice to you

Reply:
> because the reference path critic (err renamed to PathAlignCritic) by itself won't do this. If there's no "pull" it will align but not go along well, so I understand then. 
 
 +
 
> However, if you disagree and would rather 2 critics PathAlign / PathFollow, I think your name of PathFollowCritic is very apt and I can understand the reasoning for wanting them separate. I think that's an equally good choice.
 
I would prefer to decompose this. Both of them can be logically separated (PathFollow and PathAlign), and they quite big  each, so from software perspective i think it would be better not to mix them and overload parameters for one critic

Reply:
decomposed